python/GafferTest/NodeAlgoTest.py

```diff
@@ -361,6 +361,89 @@ def testBadVisitorReturnValue( self ) :
 		with self.assertRaisesRegexp( RuntimeError, "Visitor must return a bool \(True to continue, False to prune\)" ) :
 			Gaffer.NodeAlgo.visitUpstream( g["L3_1"], lambda node : None )
 
+	def __boxedVisitationGraph( self ) :
+
+		s = self.__visitationGraph()
+		b = Gaffer.Box.create( s, Gaffer.StandardSet( [ s["L1_1"] ] ) )
+		b.setName( "Box_L1_1" )
+		b = Gaffer.Box.create( s, Gaffer.StandardSet( [ s["L2_3"] ] ) )
+		b.setName( "Box_L2_3" )
+		return s
+
+	def testVisitBoxedNodesDepthFirst( self ) :
+
+		s = self.__boxedVisitationGraph()
+
+		self.assertEqual(
+			Gaffer.NodeAlgo.downstreamNodes( s["Box_L1_1"], order = Gaffer.NodeAlgo.VisitOrder.DepthFirst ),
+			[ s[ "L2_1"], s["L3_1"], s["L3_2"] ],
+		)
+
+		self.assertEqual(
+			Gaffer.NodeAlgo.downstreamNodes( s["Box_L1_1"]["L1_1"], order = Gaffer.NodeAlgo.VisitOrder.DepthFirst ),
+			[ s["Box_L1_1"], s[ "L2_1"], s["L3_1"], s["L3_2"] ],
+		)
+
+		self.assertEqual(
+			Gaffer.NodeAlgo.downstreamNodes( s["L1_2"], order = Gaffer.NodeAlgo.VisitOrder.DepthFirst ),
+			[ s["L2_2"], s["L3_2"], s["Box_L2_3"], s["Box_L2_3"]["L2_3"] ],
+		)
+
+		self.assertEqual(
+			Gaffer.NodeAlgo.upstreamNodes( s["Box_L2_3"], order = Gaffer.NodeAlgo.VisitOrder.DepthFirst ),
+			[ s[ "L1_2"] ],
+		)
+
+		self.assertEqual(
+			Gaffer.NodeAlgo.upstreamNodes( s["Box_L2_3"]["L2_3"], order = Gaffer.NodeAlgo.VisitOrder.DepthFirst ),
+			[ s["Box_L2_3"], s[ "L1_2"] ],
+		)
+
+		self.assertEqual(
+			Gaffer.NodeAlgo.upstreamNodes( s["L3_1"], order = Gaffer.NodeAlgo.VisitOrder.DepthFirst ),
+			[ s[ "L2_1"], s["Box_L1_1"], s["Box_L1_1"]["L1_1"] ],
+		)
+
+	def testVisitBoxedNodesBreadthFirst( self ) :
+
+		s = self.__boxedVisitationGraph()
+
+		self.assertEqual(
+			Gaffer.NodeAlgo.downstreamNodes( s["Box_L1_1"], order = Gaffer.NodeAlgo.VisitOrder.BreadthFirst ),
+			[ s[ "L2_1"], s["L3_1"], s["L3_2"] ],
+		)
+
+		self.assertEqual(
+			Gaffer.NodeAlgo.downstreamNodes( s["Box_L1_1"]["L1_1"], order = Gaffer.NodeAlgo.VisitOrder.BreadthFirst ),
+			[ s["Box_L1_1"], s[ "L2_1"], s["L3_1"], s["L3_2"] ],
+		)
+
+		self.assertEqual(
+			Gaffer.NodeAlgo.downstreamNodes( s["L1_2"], order = Gaffer.NodeAlgo.VisitOrder.BreadthFirst ),
+			[ s["L2_2"], s["Box_L2_3"], s["Box_L2_3"]["L2_3"], s["L3_2"] ],
+		)
+
+		self.assertEqual(
+			Gaffer.NodeAlgo.upstreamNodes( s["Box_L2_3"], order = Gaffer.NodeAlgo.VisitOrder.BreadthFirst ),
+			[ s[ "L1_2"] ],
+		)
+
+		self.assertEqual(
+			Gaffer.NodeAlgo.upstreamNodes( s["Box_L2_3"]["L2_3"], order = Gaffer.NodeAlgo.VisitOrder.BreadthFirst ),
+			[ s["Box_L2_3"], s[ "L1_2"] ],
+		)
+
+		self.assertEqual(
+			Gaffer.NodeAlgo.upstreamNodes( s["L3_1"], order = Gaffer.NodeAlgo.VisitOrder.BreadthFirst ),
+			[ s[ "L2_1"], s["Box_L1_1"], s["Box_L1_1"]["L1_1"] ],
+		)
+
+		x = Gaffer.NodeAlgo.upstreamNodes( s["L3_2"], order = Gaffer.NodeAlgo.VisitOrder.BreadthFirst )
```

Comment:
🗺️ `x` marks the spot of bonus treasure (there are also trailing commas on all the new expected results lines)

Reply:
a5ec2b7

Reply:
Excellent, excellent. I shall cancel the pending order with punctionation4u.com as stocks are looking good now.